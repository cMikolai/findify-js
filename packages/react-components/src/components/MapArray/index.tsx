import react from 'react'

export type MapCallback = (item: any, index: number, arrayLike) => any
export type KeyAccessor = (item: any) => string
export type ArrayLike = {
  map: (callback: MapCallback) => any
}
export type ReactFactory = (props: object) => react.Component
export type MapArrayProps = { array: ArrayLike, keyAccessor: KeyAccessor, factory: ReactFactory }

export default function MapArray({array, keyAccessor, factory}: MapArrayProps) {
  return array.map(item => factory({ item, key: keyAccessor(item) }))
}
