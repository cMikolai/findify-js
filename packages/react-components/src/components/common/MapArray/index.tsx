import React from 'react'

export type MapCallback = (item: any, index: number, arrayLike) => any
export type KeyAccessor = (item: any, index: number) => string
export type ArrayLike = {
  map: (callback: MapCallback) => any,
  length: number,
  slice: (from: number, to?: number) => ArrayLike
}
export type ReactFactory = (props: object) => React.Component
export type MapArrayProps = {
  array: ArrayLike,
  keyAccessor?: KeyAccessor,
  factory: ReactFactory,
  limit?: number,
  [key: string]: any
}

const defaultKeyAccessor = (item, index) => item.hashCode();

export default ({
  array,
  keyAccessor = defaultKeyAccessor,
  factory,
  limit,
  ...rest
}: MapArrayProps) => {
  const f = React.createFactory(factory);
  const res: any = Array.from(
    array.slice(0, limit || array.length)
      .map((item, index) => f({ ...rest, item, index, key: keyAccessor(item, index) }))
  );
  return res;
}
