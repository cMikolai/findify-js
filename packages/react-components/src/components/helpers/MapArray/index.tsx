import React from 'react'

export type MapCallback = (item: any, index: number, arrayLike) => any
export type KeyAccessor = (item: any, index: number) => string
export type ArrayLike = {
  map: (callback: MapCallback) => any,
  length: number,
  slice: (from: number, to?: number) => ArrayLike
}
export type ReactFactory = (props: object) => React.Component
export type MapArrayProps = { array: ArrayLike, keyAccessor: KeyAccessor, factory: ReactFactory, limit: number }

const defaultKeyAccessor = (item, index) => item.hashCode();

export default function MapArray({array, keyAccessor = defaultKeyAccessor, factory, limit}: MapArrayProps) {
  const f = React.createFactory(factory);
  return Array.from(
    array.slice(0, limit || array.length)
      .map((item, index) => f({ item, index, key: keyAccessor(item, index) }))
  )
}
