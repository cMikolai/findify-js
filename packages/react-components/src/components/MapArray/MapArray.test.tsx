import React from 'react'
import { shallow } from 'enzyme'
import MapArray from './index'
import itemMock from '../../../jest/item.mock';

const items = [
  itemMock(
    'Test Title',
    'https://picsum.photos/100/200'
  ),
  itemMock(
    'Test Title 2',
    'https://picsum.photos/100/200'
  )
]

describe('ItemCard', () => {
  it('renders correctly', () => {
    expect(shallow(
      <MapArray
        array={items}
        factory={({ item, key }) =>
          <div key={key}>{item.get('title')} {item.get('image_url')}</div>
        }
        keyAccessor={item => item.hashCode()} />)).toMatchSnapshot()
  })
})
