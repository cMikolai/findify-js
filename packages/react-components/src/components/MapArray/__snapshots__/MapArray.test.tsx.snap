// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemCard renders correctly 1`] = `
Array [
  <div
    key="Test Titlehttps://picsum.photos/100/200"
  />,
  <div
    key="Test Title 2https://picsum.photos/100/200"
  />,
]
`;
